crates/ruff_linter/src/rules/pylint/rules/invalid_bytes_return.rs

```diff
@@ -0,0 +1,77 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::ReturnStatementVisitor;
+use ruff_python_ast::visitor::Visitor;
+use ruff_python_ast::Stmt;
+use ruff_python_semantic::analyze::type_inference::{PythonType, ResolvedPythonType};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for `__bytes__` implementations that return a type other than `bytes`.
+///
+/// ## Why is this bad?
+/// The `__bytes__` method should return a `bytes` object. Returning a different
+/// type may cause unexpected behavior.
+///
+/// ## Example
+/// ```python
+/// class Foo:
+///     def __bytes__(self):
+///         return 2
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class Foo:
+///     def __bytes__(self):
+///         return b"2"
+/// ```
+///
+/// ## References
+/// - [Python documentation: The `__bytes__` method](https://docs.python.org/3/reference/datamodel.html#object.__bytes__)
+#[violation]
+pub struct InvalidBytesReturnType;
+
+impl Violation for InvalidBytesReturnType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__bytes__` does not return `bytes`")
+    }
+}
+
+/// E0308
+pub(crate) fn invalid_bytes_return(checker: &mut Checker, name: &str, body: &[Stmt]) {
+    if name != "__bytes__" {
+        return;
+    }
+
+    if !checker.semantic().current_scope().kind.is_class() {
+        return;
+    }
+
+    let returns = {
+        let mut visitor = ReturnStatementVisitor::default();
+        visitor.visit_body(body);
+        visitor.returns
+    };
+
+    for stmt in returns {
```

Comment:
Should this rule and the other invalid-*-returned rules also flag the case where the method has no return statement

```
class Test:
	def __bytes__(self):
		print("nope")
```

Reply:
I added this case as well. If this becomes the consensus, I can also add it to the already included bool (E0304) and str (E0307) cases.

Reply:
@MichaReiser As it turned out, the no-return case is not that simple: if there is just a pass/ellipsis/raise statement, then this rule should not raise. However, if there is e.g. a print statement and a raise statement, then it should raise - see the tests for examples.

Reply:
I changed this to use `is_stub` for detecting the other cases.